dev/service-registry/service-locator/src/main/scala/com/lightbend/lagom/gateway/AkkaHttpServiceGateway.scala

```diff
@@ -0,0 +1,142 @@
+          case Some(upgrade) =>
+            handleWebSocketRequest(request, newUri, upgrade)
+          case None =>
+            http.singleRequest(request.withUri(newUri).withHeaders(filterHeaders(request.headers)))
+        }
+      case NotFound(registryMap) =>
+        log.debug("Sending not found response")
+        Future.successful(renderNotFound(request, path, registryMap))
+    }
+  }
+
+  private def handleWebSocketRequest(request: HttpRequest, uri: Uri, upgrade: UpgradeToWebSocket) = {
+    log.debug("Switching to WebSocket protocol")
+    val wsUri = uri.withScheme("ws")
+    val flow = Flow.fromSinkAndSourceMat(Sink.asPublisher[Message](fanout = false), Source.asSubscriber[Message])(Keep.both)
+
+    val (responseFuture, (publisher, subscriber)) = http.singleWebSocketRequest(
+      WebSocketRequest(wsUri, extraHeaders = filterHeaders(request.headers),
+        upgrade.requestedProtocols.headOption),
+      flow
+    )
+
+    responseFuture.map {
+
+      case ValidUpgrade(response, chosenSubprotocol) =>
+        val webSocketResponse = upgrade.handleMessages(
+          Flow.fromSinkAndSource(Sink.fromSubscriber(subscriber), Source.fromPublisher(publisher)),
+          chosenSubprotocol
+        )
+        webSocketResponse.withHeaders(webSocketResponse.headers ++ filterHeaders(response.headers))
+
+      case InvalidUpgradeResponse(response, cause) =>
+        log.debug("WebSocket upgrade response was invalid: {}", cause)
+        response
+    }
+  }
+
+  private def renderNotFound(request: HttpRequest, path: String, registry: Map[String, ServiceRegistryService]): HttpResponse = {
+    import scala.collection.JavaConverters._
+    import scala.compat.java8.OptionConverters._
+    // We're reusing Play's not found error page here, which lists the routes, we need to convert the service registry
+    // to a Play router with all the acls in the documentation variable so that it can render it
+    val router = new SimpleRouter {
+      override def routes: Routes = PartialFunction.empty
+      override val documentation: Seq[(String, String, String)] = registry.toSeq.flatMap {
+        case (serviceName, service) =>
+          val call = s"Service: $serviceName (${service.uri})"
+          service.acls().asScala.map { acl =>
+            val method = acl.method.asScala.fold("*")(_.name)
+            val path = acl.pathRegex.orElse(".*")
+            (method, path, call)
+          }
+      }
+    }
+
+    val html = views.html.defaultpages.devNotFound(request.method.name, path, Some(router)).body
+    HttpResponse(
+      status = StatusCodes.NotFound,
+      entity = HttpEntity(
+        ContentTypes.`text/html(UTF-8)`,
+        html
+      )
+    )
+  }
+
+  private val HeadersToFilter = Set(
+    "Timeout-Access",
+    "Sec-WebSocket-Accept",
+    "Sec-WebSocket-Version",
+    "Sec-WebSocket-Key",
+    "UpgradeToWebSocket",
+    "Upgrade",
+    "Connection"
+  ).map(_.toLowerCase(Locale.ENGLISH))
```

Comment:
This list is probably not complete, but it's really hard to know which headers will cause an issue. There may be some that when akka-http sees it, it'll output a warning saying you shouldn't specify that header directly (the `Timeout-Access` header, which is a meta head for providing some higher level callbacks, is one example). The incompleteness of this list is the most likely source of bugs in this implementation.